### PR TITLE
TASK: Content cache adjustments

### PR DESCRIPTION
**Pull request by nezaniel**
This adds an interface for other content caches than Neos' Fusion content cache.

Also flushes caches for deleted nodes, not just their parents, in case they or their types were used in CacheTags elsewhere

**Checklist**

- [x] Code follows the PSR-2 coding style
- [x] Tests have been created, run and adjusted as needed
- [x] The PR is created against the [lowest maintained branch](https://www.neos.io/features/release-roadmap.html)
- [ ] Reviewer - PR Title is brief but complete and starts with `FEATURE|TASK|BUGFIX`
- [ ] Reviewer - The first section explains the change briefly for change-logs
- [ ] Reviewer - Breaking Changes are marked with `!!!` and have upgrade-instructions

### CODE REVIEW

File: Neos.Neos/Classes/Cache/ContentCacheFlusherInterface.php [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+use Neos\Media\Domain\Model\AssetInterface;
+
+/**
+ * The interface for services flushing content caches triggered by node changes.
+ *
+ * It is called when the projection changes: In this case, it is triggered by
+ * {@see GraphProjectorCatchUpHookForCacheFlushing} which calls this method..
+ *   This is the relevant case if publishing a workspace
+ *   - where we f.e. need to flush the cache for Live.
+ */
```
Comment (mhsdesign) [inline Neos.Neos/Classes/Cache/ContentCacheFlusherInterface.php:30]:
```suggestion
 * @internal this extension point is experimental 
 */
```

what do you say? is this interface subject to change or can we safely declare it as `@api`

Comment (nezaniel) [reply Neos.Neos/Classes/Cache/ContentCacheFlusherInterface.php:30]: internal sounds good for now

**Comment by kitsunet** [comment]
Isn't the point of this to be not internal in order for people to use this? I guess fine for a start but this is supposed to become API right?

**Review by bwaidelich — DISMISSED** [review]
Sorry for the -1, but I'm really skeptical about this (as mentioned on [Slack](https://neos-project.slack.com/archives/C3MCBK6S2/p1697883870785069)).

IMO we suffer from Interface-itis in the "old" Neos code base. I like interfaces but they should have a very clear role and in this case it is not evident to me.
i.e. What is an implementation of `flushNodeAggregate` supposed to do? (the name of this method is rather unfortunate IMO – not a new issue of this PR, but extracting it to an interface makes it more prominent)
Why are assets part of the deal and not other entities?

My main concern is, that this introduces another _global_ point of replacement by using `Objects.yaml` – but changing the `ContentCacheFlusherInterface` implementation could break implementations that rely on the current content cache behavior..

Long story short:

I think it's a good idea to allow for 3rd parties to hook into the "content cache flushing" but IMO that should not be in the form of an adapter pattern but rather some kind of extension/hook/callback.
Probably the easiest solution would be some option for the `GraphProjectorCatchUpHookForCacheFlushing` that allows for registering custom code to be executed _on a per CR basis_.

Again, sorry for stepping on the brakes once again.. I'm happy to discuss this and come up with a solution together

**Comment by kitsunet** [comment]
Good points, maybe we need an idea of projected use cases? I can see a point of getting informed when content caches are flushed and a case of being informed when the projections changed, but both could be done differntly, but overwriting the cache flushing in itself? Thinking about it I agree this might be the "wrong" extension point.

**Comment by nezaniel** [comment]
@bwaidelich yes, let's do this together... I have quite some ideas for improvements, including CacheTag VOs, Unit tests etc.

**Review by kitsunet — APPROVED** [review]
LGTM, but yes, an internal interface won't help much, so either way, if we want to merge this, I am fine, but I could also see room for something nicer.

**Comment by mhsdesign** [comment]
I think a interface like this to replace cache flushing is also bad practice, if we were to make this an extension point we would need some kind of chain where we can hook multiple flushers, as the `ContentCacheFlusher` is probably necessary.

I know that having an interface is more fancy but still, objects yaml exists and the `ContentCacheFlusher` is not final. So for your experimental usecase i would suggest to go down the road of extending the flusher and adding functionality and then for every place it is used use `Objects.yaml` to use your implementation as constructor argument.

That would make in my head much more sense at this stage. We as core devs have a little perk of being able to easily add what we want...

File: Neos.Neos/Classes/Fusion/Cache/GraphProjectorCatchUpHookForCacheFlushing.php [hunk 2]
```diff
@@ -145,6 +146,11 @@ public function onBeforeEvent(EventInterface $eventInstance, EventEnvelope $even
                 $eventInstance->getNodeAggregateId()
             );
             if ($nodeAggregate) {
+                $this->scheduleCacheFlushJobForNodeAggregate(
+                    $this->contentRepository,
+                    $nodeAggregate->contentStreamId,
+                    $nodeAggregate->nodeAggregateId
+                );
```
Comment (mhsdesign) [inline Neos.Neos/Classes/Fusion/Cache/GraphProjectorCatchUpHookForCacheFlushing.php:153]: still this part seems to be a bugfix that we want to preserve?

**Comment by dlubitz** [comment]
@nezaniel Is the interface still relevant?

> flushes caches for deleted nodes, not just their parents

This is fixed by #5175

**Comment by nezaniel** [comment]
I think this can be closed, I've opted for AOP to solve this in the ComponentView for now.
